frontend/src/actions/paginationActions.js

```diff
@@ -1,39 +1,82 @@
-import { setTelcoPage, setTelcoPageOptions } from "./telcoActions";
+import {
+  fetchQuayJobsData,
+  setQuayOffset,
+  setQuayPage,
+  setQuayPageOptions,
+} from "./quayActions";
+import {
+  fetchTelcoJobsData,
+  setTelcoOffset,
+  setTelcoPage,
+  setTelcoPageOptions,
+} from "./telcoActions";
+
 export const setPage = (newPage, currType) => (dispatch) => {
-  if (currType === "cpt") {
-    dispatch(setCPTPage(newPage));
-  } else if (currType === "ocp") {
-    dispatch(setOCPPage(newPage));
-  } else if (currType === "quay") {
-    dispatch(setQuayPage(newPage));
-  } else if (currType === "telco") {
-    dispatch(setTelcoPage(newPage));
-  }
+  const actions = {
+    cpt: setCPTPage,
+    ocp: setOCPPage,
+    quay: setQuayPage,
+    telco: setTelcoPage,
+  };
+  dispatch(actions[currType](newPage));
 };
 
 export const setPageOptions = (newPage, newPerPage, currType) => (dispatch) => {
-  if (currType === "cpt") {
-    dispatch(setCPTPageOptions(newPage, newPerPage));
-  } else if (currType === "ocp") {
-    dispatch(setOCPPageOptions(newPage, newPerPage));
-  } else if (currType === "quay") {
-    dispatch(setQuayPageOptions(newPage, newPerPage));
-  } else if (currType === "telco") {
-    dispatch(setTelcoPageOptions(newPage, newPerPage));
-  }
+  const actions = {
+    cpt: setCPTPageOptions,
+    ocp: setOCPPageOptions,
+    quay: setQuayPageOptions,
+    telco: setTelcoPageOptions,
+  };
+  dispatch(actions[currType](newPage, newPerPage));
+};
+
+const calculateOffset = (pageNumber, itemsPerPage) => {
+  return (pageNumber - 1) * itemsPerPage;
 };
 
-export const sliceTableRows = (startIdx, endIdx, currType) => (dispatch) => {
-  if (currType === "cpt") {
-    dispatch(sliceCPTTableRows(startIdx, endIdx));
-  } else if (currType === "ocp") {
-    dispatch(sliceOCPTableRows(startIdx, endIdx));
+export const checkTableData = (newPage, currType) => (dispatch, getState) => {
+  const { results, totalJobs, perPage, page } = getState()[currType];
+  const fetchActions = {
+    ocp: fetchOCPJobs,
+    quay: fetchQuayJobsData,
+    telco: fetchTelcoJobsData,
+  };
+  const offsetActions = {
+    ocp: setOCPOffset,
+    quay: setQuayOffset,
+    telco: setTelcoOffset,
+  };
+  const hasPageData = results.length >= newPage * perPage;
+  const offset = calculateOffset(newPage, perPage);
+  if (results.length < totalJobs && !hasPageData) {
+    if (currType === "cpt") {
+      const startIdx = (page - 1) * perPage;
+      const endIdx = startIdx + perPage - 1;
+      if (results[startIdx] === undefined || results[endIdx] === undefined) {
+        dispatch(fetchOCPJobsData());
+      }
+    } else {
+      dispatch(offsetActions[currType](offset));
+      dispatch(fetchActions[currType]());
+    }
+  } else {
+    if (currType === "cpt") {
+      const startIdx = (page - 1) * perPage;
+      const endIdx = startIdx + perPage - 1;
+      dispatch(sliceCPTTableRows(startIdx, endIdx));
+    }
```

Comment:
Because each tab/component does really different things, and therefore the "common" components have distinct cases. It's not a great design. For example, we might be better off providing common base classes that get extended by components to handle behavioral differences ... assuming those differences can't simply be resolved.

For now, although the code has been pushed and stretched a bit around the edges, this aspect of the logic isn't new and we really need to push these PRs through and move on so let's defer the analysis of possible deep architectural cleanup.